### PR TITLE
Fix #161: Restore Shields local & global stats

### LINKED ISSUE #161: Add Shield Stats
Currently no shield stats are persisted or accumulating.

Known affected Files: `BraveShieldState.swift`, `ShieldsViewController.swift`, `BraveShieldStatsView.swift`, `BrowserViewController.swift`

1. Move the stat counts storage to `Preferences` (which may mean bringing the base implementation of `Preferences` to `BraveShared`, or pulling counts out entirely from `BraveShieldState` to `Client`)

2. Add the **local** counts back to the shields panel (`ShieldsViewController.init(url: URL, shieldBlockStats: ShieldBlockedStats?)`). These need to be updated as shields are updated. This may be blocked on other shield related items and require a new ticket down the road

3. Add the counts back to the home panel for new tabs in `BraveShieldStatsView`

4. Also looks like theres something to do in Domain.swift: setBraveShield(forDomain:state:context:) and loadShieldsIntoMemory()

### PR DESCRIPTION
**Pull request by kylehickinson**
## Pull Request Checklist

- [x] My patch has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!`
- [ ] I have updated the *Unit Tests* to cover new or changed functionality
- [ ] I have updated the *UI Tests* to cover new or changed functionality
- [ ] I have made sure that localizable strings use `NSLocalizableString()`

## Screenshots

_None included_

## Notes for testing this patch

HTTPS Upgrades still display in the UI, this will be removed in a later ticket (#404)

### CODE REVIEW

File: Client/Frontend/ContentBlocker/ContentBlockerHelper.swift [hunk 1]
```diff
@@ -317,16 +285,10 @@ extension ContentBlockerHelper {
     }
 
     static func isTrackingProtectionActive(tabManager: TabManager) -> Bool {
-        guard let selectedTab = tabManager.selectedTab, let blocker = selectedTab.contentBlocker as? ContentBlockerHelper else {
+        guard let selectedTab = tabManager.selectedTab, selectedTab.contentBlocker != nil else {
```
Comment (jhreis) [inline Client/Frontend/ContentBlocker/ContentBlockerHelper.swift]: Not even sure this is needed, but no need to unwrap `selectedTab` first:

`return tabManager.selectedTab?.contentBlocker != nil`

However, if `contentBlocker` is made a lazy var (see previous comment), then this function is completely unnecessary.

File: BraveShared/Preferences.swift [hunk 2]
```diff
@@ -39,6 +39,15 @@ extension Preferences {
         static let referralCode = Option<String?>(key: "urp.referral.code", default: nil)
         static let referralCodeDeleteDate = Option<TimeInterval?>(key: "urp.referral.delete-date", default: nil)
     }
+    final class BlockStats {
+        static let adsCount = Option<Int>(key: "stats.adblock", default: 0)
```
Comment (jhreis) [inline BraveShared/Preferences.swift:43]: Are these migrated from 1.6?

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 3]
```diff
@@ -124,8 +124,7 @@ class BrowserViewController: UIViewController {
 
     let downloadQueue = DownloadQueue()
     
-    fileprivate let shieldBlockStats = ShieldBlockedStats()
-    fileprivate var contentBlockListDeferred: Deferred<()>?
+    fileprivate var contentBlockListDeferred: Deferred<((), ())>?
```
Comment (jhreis) [inline Client/Frontend/Browser/BrowserViewController.swift:127]: Why this change?

File: Client/Frontend/Browser/Tab.swift [hunk 4]
```diff
@@ -98,21 +98,19 @@ class Tab: NSObject {
...
             _noImageMode = newValue
-            let helper = (contentBlocker as? ContentBlockerHelper)
-            helper?.noImageMode(enabled: _noImageMode)
+            contentBlocker?.noImageMode(enabled: _noImageMode)
         }
     }
 
     // There is no 'available macro' on props, we currently just need to store ownership.
-    var contentBlocker: AnyObject?
+    var contentBlocker: ContentBlockerHelper?
```
Comment (jhreis) [inline Client/Frontend/Browser/Tab.swift]: Can we make this a `lazy var` to remove the optionality? I think that would simplify a number of things.

File: Client/Frontend/Browser/TabManager.swift [hunk 5]
```diff
@@ -858,7 +858,7 @@ extension TabManager: WKNavigationDelegate {
     func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
         UIApplication.shared.isNetworkActivityIndicatorVisible = true
 
-        if let tab = self[webView], let blocker = tab.contentBlocker as? ContentBlockerHelper {
+        if let tab = self[webView], let blocker = tab.contentBlocker {
```
Comment (jhreis) [inline Client/Frontend/Browser/TabManager.swift]: Is `tab` even being used?

`tab.contentBlocker?.clearPageStats()`?

File: Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift [hunk 6]
```diff
@@ -33,31 +34,57 @@ struct TPPageStats {
...
-    // TODO: 161, if this is called regardless of whitelist conditions then it will need `mainFrameURL`
-    // to know if the main frame is blocked (as opposed to solely the requested URL)
-    func isBlocked(url: URL) -> Deferred<BlocklistName?> {
+    func isBlocked(url: URL, domain: Domain, resourceType: TPStatsResourceType? = nil) -> Deferred<BlocklistName?> {
         let deferred = Deferred<BlocklistName?>()
-
-        // TODO: 161, use `urlIsInList` to identify if `url` is inside a blocklist
         
-        deferred.fill(nil)
+        guard let blockLists = blockLists, let host = url.host, !host.isEmpty else {
```
Comment (jhreis) [inline Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift:50]: Are any of these even used?

File: Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift [hunk 7]
```diff
@@ -33,31 +34,57 @@ struct TPPageStats {
...
+                        return
+                    }
+                case .image:
+                    if enabledLists.contains(.image) {
+                        deferred.fill(.image)
+                        return
+                    }
+                }
+            }
+            deferred.fill(blockLists.urlIsInList(url).flatMap { return enabledLists.contains($0) ? $0 : nil })
```
Comment (jhreis) [inline Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift]: This seems kind of funky. So `urlIsInList` returns a list and then we see if that list is in `enabledLists`?

File: Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift [hunk 8]
```diff
@@ -176,7 +201,7 @@ fileprivate class TPStatsBlocklists {
         }
     }
 
-    func urlIsInList(_ url: URL, whitelistedDomains: [NSRegularExpression]) -> BlocklistName? {
+    func urlIsInList(_ url: URL) -> BlocklistName? {
```
Comment (jhreis) [inline Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift]: I think this may need to be renamed, kind of weird.

**Review by jhreis — CHANGES_REQUESTED** [review]

File: BraveShared/Preferences.swift (hunk 2 above)
Comment (kylehickinson) [reply BraveShared/Preferences.swift:43]: Yep the 5 that are saved in 1.6 are migrated

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 3 above)
Comment (kylehickinson) [reply Client/Frontend/Browser/BrowserViewController.swift:127]: I have the rule parsing for FF's block list (for checking if a url is being blocked). You can see the in compileLists:

```swift
static func compileLists() -> Deferred<((), ())> {
    let statsList = TPStatsBlocklistChecker.shared.startup()
    let compileList = BlocklistName.compileAll(ruleStore: ruleStore)
    return statsList.both(compileList)
}
```

The deferred library lets me do 2 parallel tasks and have the deferred trigger when both finish (similar to dispatch group)

File: Client/Frontend/Browser/TabManager.swift (hunk 5 above)
Comment (kylehickinson) [reply Client/Frontend/Browser/TabManager.swift]: Tab's used below yeah

File: Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift (hunk 7 above)
Comment (kylehickinson) [reply Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift]: Yeah this is how it was but I agree its funky… 

`blockLists` (a `TPStatsBlocklists?`) calls `urlIsInList` which returns the `BlocklistName?` that url belongs too, then we fill `deferred` with that name or nil if that blacklist it found it in isn't currently enabled.

```swift
if let blocklist = blockLists.urlIsInList(url), enabledLists.contains(foundList) {
    deferred.fill(blocklist)
} else {
    deferred.fill(nil)
}
```

This is the easier to understand but more lines version of it

File: Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift (hunk 6 above)
Comment (kylehickinson) [reply Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift:50]: `blockLists` is, `host` is just used for the `!host.isEmpty` part

File: Client/Frontend/Browser/Tab.swift [hunk 9]
```diff
@@ -98,21 +98,19 @@ class Tab: NSObject {
...
             _noImageMode = newValue
-            let helper = (contentBlocker as? ContentBlockerHelper)
-            helper?.noImageMode(enabled: _noImageMode)
+            contentBlocker.noImageMode(enabled: _noImageMode)
         }
     }
 
     // There is no 'available macro' on props, we currently just need to store ownership.
-    var contentBlocker: AnyObject?
+    lazy var contentBlocker = ContentBlockerHelper(tab: self)
```
Comment (jhreis) [inline Client/Frontend/Browser/Tab.swift:113]: Oh, haven't seen this syntax before, nice.

File: Client/Frontend/ContentBlocker/ContentBlockerHelper.swift [hunk 10]
```diff
@@ -177,29 +176,7 @@ class ContentBlockerHelper {
...
-        guard isEnabled else {
-            return .Disabled
-        }
-        if stats.total == 0 {
-            // TODO: 161, _may_ need to handle whitelisted situations here
-            return .NoBlockedURLs
-        } else {
-            return .Blocking
-        }
+        return tab != nil
```
Comment (jhreis) [inline Client/Frontend/ContentBlocker/ContentBlockerHelper.swift]: Maybe just:

`return isUserEnabled == true || tab != nil`

File: Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift [hunk 11]
```diff
@@ -176,7 +205,7 @@ fileprivate class TPStatsBlocklists {
         }
     }
 
-    func urlIsInList(_ url: URL, whitelistedDomains: [NSRegularExpression]) -> BlocklistName? {
+    func blocklistForURL(_ url: URL) -> BlocklistName? {
```
Comment (jhreis) [inline Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift:208]: So much better.

File: Client/Frontend/Shields/ShieldsViewController.swift [hunk 12]
```diff
@@ -8,41 +8,45 @@ import Shared
...
+    private var statsUpdateObservable: AnyObject?
+    
     /// Create with an initial URL and block stats (or nil if you are not on any web page)
-    init(url: URL?, shieldBlockStats: ShieldBlockedStats) {
-        self.url = url
-        self.shieldBlockStats = shieldBlockStats
+    init(tab: Tab) {
+        self.tab = tab
+        self.url = tab.url
+        self.shieldBlockStats = tab.contentBlocker.stats
```
Comment (jhreis) [inline Client/Frontend/Shields/ShieldsViewController.swift]: If we are retaining `tab`, is there any point in storing these as additional properties? It also then requires keeping it updated, as in [here](https://github.com/brave/brave-ios/pull/403/files#diff-ed29423ac71a29faa319243941d34b93R48)

File: Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/TrackingProtectionStats.js [hunk 13]
```diff
@@ -35,20 +35,20 @@ function install() {
     }
   })
 
-  function sendMessage(url) {
+  function sendMessage(url, resourceType) {
     if (url) {
-      webkit.messageHandlers.trackingProtectionStats.postMessage({ url: url });
+      webkit.messageHandlers.trackingProtectionStats.postMessage({ url: url, resourceType: resourceType === undefined ? null : resourceType });
```
Comment (jhreis) [inline Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/TrackingProtectionStats.js:40]: Should we have a resource type `other` maybe?

**Review by jhreis — CHANGES_REQUESTED** [review]

File: Client/Frontend/ContentBlocker/ContentBlockerHelper.swift (hunk 10 above)
Comment (kylehickinson) [reply Client/Frontend/ContentBlocker/ContentBlockerHelper.swift]: Unfortunately this isn't the same logic then. If `isUserEnabled` is non-nil it can still return its value (even if its false)

File: Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/TrackingProtectionStats.js (hunk 13 above)
Comment (kylehickinson) [reply Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/TrackingProtectionStats.js:40]: I considered it but in the end it doesn't make much sense because if we don't know the resource type, then we aren't going to track it. (`other` would just be a synonym for `nil` at that point)

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 3 above)
Comment (jhreis) [reply Client/Frontend/Browser/BrowserViewController.swift:127]: I was assuming that would be `-> (Deferred<()>, Deferred<()>)`, but this is much better!

File: Client/Frontend/ContentBlocker/ContentBlockerHelper.swift (hunk 10 above)
Comment (jhreis) [reply Client/Frontend/ContentBlocker/ContentBlockerHelper.swift]: Ah yes, how about:

`return isUserEnabled ?? tab != nil`

?!

**Review by jhreis — APPROVED** [review]
